dough-inventories/src/main/java/io/github/bakedlibs/dough/inventory/factory/MenuListener.java

```diff
@@ -0,0 +1,98 @@
+package io.github.bakedlibs.dough.inventory.factory;
+
+import java.util.logging.Level;
+
+import javax.annotation.Nonnull;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.Event;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.inventory.InventoryClickEvent;
+import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.InventoryHolder;
+
+import io.github.bakedlibs.dough.inventory.Menu;
+import io.github.bakedlibs.dough.inventory.SlotGroup;
+import io.github.bakedlibs.dough.inventory.handlers.MenuClickHandler;
+import io.github.bakedlibs.dough.inventory.payloads.MenuPayloads;
+
+/**
+ * The {@link MenuListener} is responsible for handling any
+ * {@link Event} related to our {@link Menu}s.
+ * <p>
+ * It is registered by a {@link MenuFactory} and there should only
+ * be one {@link MenuListener} per {@link MenuFactory}.
+ * 
+ * @author TheBusyBiscuit
+ *
+ */
+class MenuListener implements Listener {
+
+    /**
+     * Our {@link MenuFactory} instance.
+     */
+    private final MenuFactory factory;
+
+    /**
+     * This constructs a new {@link MenuListener} for the given
+     * {@link MenuFactory}.
+     * 
+     * @param factory
+     *            Our {@link MenuFactory} instance
+     */
+    MenuListener(@Nonnull MenuFactory factory) {
+        this.factory = factory;
+    }
+
+    /**
+     * This returns the {@link MenuFactory} which instantiated and
+     * registered this {@link MenuListener}.
+     * 
+     * @return The {@link MenuFactory}
+     */
+    public @Nonnull MenuFactory getFactory() {
+        return factory;
+    }
+
+    /**
+     * Here we listen for the {@link InventoryClickEvent}.
+     * This event is fired whenever a {@link Player} clicks a slot
+     * in an {@link Inventory}. We only care for events which happen
+     * within our {@link Menu} though.
+     * 
+     * @param e
+     *            The {@link InventoryClickEvent} which was fired
+     */
+    @EventHandler
+    public void onClick(InventoryClickEvent e) {
+        InventoryHolder holder = e.getInventory().getHolder();
+
+        // Check if the Inventory is a Menu
+        if (holder instanceof Menu) {
+            Menu inv = (Menu) holder;
+
+            try {
```

Comment:
Wouldn't this fail to stop shift-clicking into filled slots? Let's say I have an "apples" group which is not interactable and filled with itemtype APPLE and I shift click a stack of apples in my inventories, the if clause is false so the event is never checked and apples get stacked into the first "apples" slots in the menu

Reply:
Might be a good idea to add something of the sort to TestMenuClicking too, just in case